scipy/sparse/lil.py

```diff
@@ -111,6 +115,29 @@ def __init__(self, arg1, shape=None, dtype=None, copy=False):
                 for i in range(M):
                     self.rows[i] = []
                     self.data[i] = []
+            elif len(arg1) == 2:
+                # (data, rows) format
+                (data, rows) = arg1
+                rows_len = len(rows)
+
+                if len(data) != rows_len:
+                    raise ValueError('data and rows should have same length')
+                
+                if shape is None:
+                    raise ValueError('shape parameter is required with (data, rows) format')
+
+                self._shape = check_shape(shape)
+                
+                if shape[0] != rows_len:
+                    raise ValueError('shape should match with length of data and rows') 
+
+                self.data = np.empty(rows_len, dtype=object)
+                self.rows = np.empty(rows_len, dtype=object)
+
+                for i in range(rows_len):
```

Comment:
I think it may be simpler to do:

```
self.data = np.array([list(d) for d in data], dtype=object)
self.rows = np.array([list(r) for r in rows], dtype=object)
```

Reply:
Initially, I tried something similar to your suggestion but it doesn't work as expected, the problem is that `np.array` constructor converts the list of lists with all numerical values to 2 ndim array.

```python
>>> data1 = [[1], [2], [3]]
>>> rows1 = [[0], [1], [2]]
>>> B = sps.lil_matrix((data1, rows1), shape=(3, 3))
>>> B.data
array([list([1]), list([2]), list([3])], dtype=object)         # we need this
```
But with passing list of lists to `np.array` we get this
```python
>>> B.data
array([[1],
       [2],
       [3]], dtype=object)```